[#12] 유저 프로필 등록 기능 구현

cold-pumpkin opened the pull request:
Closes #12 
### 유저 프로필 등록 기능

- [X] 프로필 이미지 저장용 스토리지로 AWS S3 도입
- [X] 유저 프로필 정보 저장을 위한 DB 스키마 설계
- [X] 프로필 이미지 및 상태 메시지 등록 서비스 구현
- [X] 프로필 등록 테스트 코드 작성

cold-pumpkin committed:
[#12] AWS S3 연동 작업

- 프로필 이미지 스토리지로 AWS S3를 사용하기 위해 AWS SDK for Java 2.0
  의존성 추가
- 이미지 업로드 기능을 갖춘 업로더 인터페이스 및 구현 클래스 추가

cold-pumpkin committed:
[#12] 유저 프로필 DB 스키마 및 프로필 정보 저장 매퍼 추가

cold-pumpkin committed:
[#12] 불필요한 클래스 제거

cold-pumpkin committed:
[#12] 유저 프로필 클래스 추가

cold-pumpkin committed:
[#12] 유저 프로필 등록 서비스 구현

cold-pumpkin committed:
[#12] 유저 프로필 등록 컨트롤러 구현

cold-pumpkin committed:
[#12] 유저 프로필 등록 테스트 코드 작성

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/aws/S3Uploader.java:
```diff
@@ -0,0 +1,57 @@
+
+@Slf4j
+@RequiredArgsConstructor
+public final class S3Uploader implements ImageUploader {
```
1. S3에 이미지를 업로드 시키는건지 파일을 업로드시키는지 명확하지 않습니다~
2. 이 클래스 내용을 봤을때 딱 이미지에 한정지어서 업로드시키는건 아닌 것 같은데 그냥 `FileUploader`로 인터페이스 이름을 지어줘도 될 것 같아요~

cold-pumpkin replied:
넵 우선 `FileUploader` 인터페이스로 네이밍 해두되, 추후 프로필 등록시에는 이미지만 등록할 수 있도록 로직을 추가해야 할 것 같습니다!

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/controller/user/ProfileRestController.java:
```diff
@@ -0,0 +1,24 @@
+
+    private final ProfileService profileService;
+
+    @PostMapping(path = "/profile")
```
`POST /my-infos/profile` 이렇게 가보면 어떨까요?

cold-pumpkin replied:
넵 이 컨트롤러는 가입한 유저들의 세부 정보들 중 프로필에 관련한 요청들을 처리하기 때문에 `POST /my-infos/profile`이 의미를 더 잘 전달할 수 있을 것 같습니다~

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/service/user/ProfileService.java:
```diff
@@ -0,0 +1,53 @@
+        try {
+            profileImageUrl = imageUploader.upload(profileImage);
+        } catch (S3Exception | IOException e) {
+            log.error("Image upload failed : {}", e.getMessage());
```
여기서는 예외를 던져주는게 나을 것 같은데 로그만 찍고 정상적인 흐름으로 진행하시는 이유가 있으실까요?

cold-pumpkin replied:
아 넵 개발 예외 던지도록 한다는걸 로그만 찍어놓고 놓쳤네요!

f-lab-dev commented on src/test/java/flab/project/sharemyhobby/aws/S3UploaderTest.java:
```diff
@@ -0,0 +1,65 @@
+//import com.amazonaws.auth.BasicAWSCredentials;
+//import com.amazonaws.regions.Regions;
+//import com.amazonaws.services.s3.AmazonS3;
+//import com.amazonaws.services.s3.AmazonS3ClientBuilder;
```
주석은 빼고 커밋해주시는게 좋습니다~ develop에는 최대한 완벽한 코드만 들어가야합니다

cold-pumpkin replied:
앗 주석 삭제하겠습니다~

f-lab-dev commented on src/test/java/flab/project/sharemyhobby/service/user/ProfileServiceTest.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Test
+    @DisplayName("프로필을 등록하면 DB에 저장된 프로필 정보를 리턴한다")
```
실패하는 테스트도 작성해주시면 좋을 것 같아요~ 업로드 도중에 예외가 발생한다거나요

cold-pumpkin replied:
넵 실패 케이스도 작성해보겠습니다~

cold-pumpkin committed:
[#12] 리뷰반영: 인터페이스 네이밍 변경

cold-pumpkin committed:
[#12] 리뷰반영: 유저 프로필 관련 API URL 수정

cold-pumpkin committed:
[#12] 리뷰반영: 불필요한 주석 삭제

cold-pumpkin committed:
[#12] 리뷰반영: 프로필 이미지 업로드 실패 시 예외 처리

cold-pumpkin committed:
[#12] 리뷰반영: 이미지 업로드 실패 테스트 코드 작성

- 빠른 테스트 실행을 위해 스프링 컨테이너에 의존하지 않는 단위 테스트로 변경
- 예외 강제 발생 코드의 영향이 전파되지 않도록 테스트 인스턴스 라이프사이클을
  메소드 단위로 변경

cold-pumpkin committed:
[#15] AWS S3 이미지 저장 경로 수정

f-lab-dev commented on src/test/java/flab/project/sharemyhobby/service/user/ProfileServiceTest.java:
```diff
@@ -41,24 +49,35 @@ private MultipartFile getImageFile(String fileName) throws IOException {
+    }
+
+    @Test
+    @DisplayName("프로필을 등록을 실패하면 FileUploadException 발생시킨다")
```
프로필 등록이 실패하는것은 무조건 파일업로드 실패 때문일까요?

cold-pumpkin replied:
아하 이미지 파일을 읽어오는 과정에서도 `IOException`으로 인해 실패할 수 있습니다! 
1. 업로드 중 실패하는 케이스
2.  파일 읽기 중 실패하는 케이스

위 두개로 나누고 테스트 메소드 명도 변경해 보겠습니다~

cold-pumpkin committed:
Merge pull request #16 from f-lab-edu/feature/#15

[#15] AWS S3 이미지 저장 경로 수정

cold-pumpkin committed:
[#12] 리뷰반영: 프로필 업데이트 실패 테스트 코드 추가

f-lab-dev commented on src/test/java/flab/project/sharemyhobby/service/user/ProfileServiceTest.java:
```diff
@@ -66,9 +67,24 @@ void testRegisterProfileAndReturnProfileInfo() throws IOException {
+    }
+
+    @Test
+    @DisplayName("이미지 파일 읽기 실패 시 FileUploadException 발생시킨다")
```
데이터베이스에서도 문제가 발생할 수 있지 않을까요? 프로필 등록을 실패하는 케이스는 파일과 관련된 원인뿐만은 아닐 것 같습니다~

cold-pumpkin replied:
프로필 등록시 이미 유저의 프로필 데이터가 들어가 있는 경우 `insert`쿼리 실행 중 에러가 날 수 있습니다! 
이런 경우 예외처리 하는 로직을 빼먹어서 서비스에 추가하고, 테스트 코드 작성했습니다.
혹시 의도하신 것이 맞을까요~?

cold-pumpkin committed:
[#12] 해당 유저 프로필 데이터 존재여부 체크 로직 추가

cold-pumpkin committed:
[#12] 프로필 중복 로직 테스트 코드 추가

f-lab-dev commented on src/main/java/flab/project/sharemyhobby/service/user/ProfileService.java:
```diff
@@ -33,6 +34,9 @@ public Profile registerProfile(Long userId, MultipartFile profileImage, String s
                     .statusMessage(statusMessage)
                     .build();
 
+        if (profileMapper.isDuplicate(userId))
```
유니크가 걸려있으면 이거는 굳이 검사할 필요가 없지 않을까요?

cold-pumpkin replied:
@f-lab-dev `profile` 테이블의 `user_id`를 유니크 키로 지정하고 `isDuplicate()`를 삭제한 다음 바로 `saveProfile()`을 호출하면 중복으로 프로필이 insert될 때 `SQLIntegrityConstraintViolcationException`이 발생하는데요, 그럼 이 경우 `saveProfile()`를 try-catch를 묶고 `DuplicateProfileException`을 던지는 방식으로 구현하는게 좋을까요?

f-lab-dev replied:
네 그렇게 전환시켜줘도 될 것 같네요~

cold-pumpkin replied:
넵 한번 수정해서 올려보겠습니다~

cold-pumpkin committed:
[#12] 리뷰반영: 유저 프로필 존재여부 체크를 DB 조회 방식에서 유니크키 추가로 변경

cold-pumpkin committed:
[#20] 취미 관련 테이블 스키마 설계

- 취미 카테고리 및 취미 테이블 스키마 정의
- 유저가 선택한 관심 취미 테이블 스키마 정의
- 기본 취미 정보 입력

cold-pumpkin committed:
[#12] Conflict 해결: sql 파일 위치 변경
